feat: maker keyword recognizer tutorial (#164)
spokestack/spokestack-website: content/blog/keyword-recognizer-python-tutorial/index.md

---
title: A Swear Jar in 100 Lines of Python
date: '2021-03-25'
description: Use Spokestack's keyword recognizer model and Python library to help save up for a rainy day.
author: josh
tags: Keyword Recognition, Maker, Python, Tutorial
draft: true
---

![swear jar](images/swear_jar.jpg 'swear jar')

In this tutorial, we're going to use Spokestack's AutoML training tool to create a keyword recognizer model and use it, along with the [spokestack-python](https://github.com/spokestack/spokestack-python) library, to make a digital swear jar. The full code is [on GitHub](https://github.com/spokestack/swear-jar-python) if you'd like to run it for yourself.

The concept of a swear jar is simple: you have a list of words or short phrases that you'd like to stop saying, or stop _someone else_ from saying. Every time one of those words does slip out, you drop a coin in a jar. When the jar's full, you take it to the bank and open a savings account to spend when you're better behaved.

> We should note that if swearing's not your thing, this concept also lets you reward yourself for saying nice things—call it a "compliment jar". We wouldn't want to limit this tutorial to us reprobates.

Reviewer: this joke lands

Anyway, the first thing that might come to mind for this use case is a [wakeword detector](/docs/concepts/wake-word): you don't want your app running ASR all day, continuously streaming a pipe of data to wherever; you just want it to notice when you've said certain things. But there's a wrinkle: if movie ratings have taught us anything, it's that certain words are just . . . ickier than others. We'd like to make it more expensive to say _\<insert epithet here\>_ than, say, _\<insert milder epithet here\>_. (You didn't think we were actually going to suggest profanity for your list, did you? This is a family-friendly tech site we're running here.)

A [keyword recognizer](/docs/concepts/keywords) is perfect for this. Trained on a small set of utterances, it acts as a wakeword recognizer and ASR in one, letting you know which utterance it heard without activating a cloud service in the process—your bad habits stay on your device forever.

Now that we know what we're doing and why, let's get to _doing_ it.

## Creating a Model

> _**Note**: You'll need a Spokestack Maker account to follow along with model creation. Running the sample app is totally free, so go ahead and take it for a spin first. Keep in mind that the models provided with the sample code are [personal models](/docs/concepts/keywords#personal-keyword), so performance with your voice may vary. They were trained with a relatively deep voice._

Reviewer: ```suggestion
_**Note**: You'll need a Spokestack Maker account to follow along with model creation. Running the sample app is totally free, so go ahead and take it for a spin first. Keep in mind that the models provided with the sample code are [personal models](/docs/concepts/keywords#personal-keyword), so [performance with your voice may vary](what-are-personal-ai-models). They were trained with a relatively deep voice._
```

If this is your first time using Spokestack, you'll need [an account](/create). If you have a free account, you'll need to [upgrade to the Maker tier](/account/settings#billing). Once that's taken care of, head over to the [keyword tool](/account/keyword). Click the "New Model" and name it whatever you like. We're going with "swear jar" here, saving our creativity for the actual list of words. Use "add keyword" button until you've accumulated enough to break (or reinforce) your habit.

> While adding keywords, you'll notice that you also have the option to add multiple utterances to each keyword. Keywords and utterances interact like this: the keyword is the text that will be returned to your app when the model recognizes any utterance listed under it. So if one of your swears is, say, "beef", but you also want to stop saying "beefsteak", you might group those together by making them separate utterances under the keyword "beef".

Reviewer: ```suggestion
If this is your first time using Spokestack, you'll need [an account](/create). If you have a free account, you'll need to [upgrade to the Maker tier](/account/settings#billing). Once that's taken care of, head over to the [keyword tool](/account/keyword). Click the "New Model" and name it whatever you like. We're going with "swear jar" here, saving our creativity for the actual list of words. 

Speaking of, now's the moment you've been waiting for:  channel your inner George Carlin and click that "add keyword" button until you've accumulated enough to break (or reinforce) your habit.

While adding keywords, you'll notice that you also have the option to add multiple utterances to each keyword. Keywords and utterances interact like this: the keyword is the text that will be returned to your app when the model recognizes any utterance listed under it. So if one of your swears is, say, "beef", but you also want to stop saying "beefsteak", you might group those together by making them separate utterances under the keyword "beef".
```

```

The other piece of our puzzle is the app's reaction when it hears a word on our list. For that we'll use the pipeline event we registered above:

```python
from spokestack.io.pyaudio import PyAudioOutput

AUDIO_DIR = "audio"
OUTPUT = PyAudioOutput(num_channels=1, sample_rate=44100)

def swear_heard(context):
  value = WORDS.get(context.transcript)
  if value:
    sound_path = os.path.join(AUDIO_DIR, f"{value}.wav")
    with wave.open(sound_path, 'rb') as coin:
      frame = coin.readframes(1024)
      while frame != '':
        OUTPUT.write(frame)
    print(f"Uh-oh! I heard {context.transcript}. That'll be {value} cents.")
```

Notice that this function was registered to listen to the `recognize` event, hinting at the fact that the keyword recognizer is acting as our ASR here. If we were using a wake word detector instead, it would fire the `activate` event, and we'd have to wait for `recognize` from a separate ASR to get a meaningful value in `context.transcipt.`

Thanks to some pre-recorded and cleverly named audio files, we're able to not only print a scolding message to the terminal, but play the sound of a hard-earned coin dropping into the jar.

> This code could be modified to keep a running total of the amount deposited throughout the day, but that feature is left as an exercise to the reader.

Reviewer: Good idea—maybe this is a `Go Deeper` section prompt?

Author: As in, you think I should add that to the tutorial? I see that as "just code"—sort of out of scope for what we're doing here; including it would make the tutorial longer/more intimidating for little gain.

Reviewer: ```suggestion
## Go Deeper
- Keep a running total of the amount deposited throughout the day
- Grab a housemate and train using both your voices
- ???
```
